packages/lit-ssr/src/test/integration/client/basic_test.ts

```diff
@@ -0,0 +1,331 @@
+/**
+ * @license
+ * Copyright (c) 2020 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import '@open-wc/testing';
+
+import {tests} from '../tests/basic.js';
+import {render} from 'lit-html';
+import {hydrate} from 'lit-html/hydrate.js';
+import {hydrateShadowRoots} from 'template-shadowroot/template-shadowroot.js';
+import {SSRExpectedHTML} from '../tests/ssr-test.js';
+
+const assert = chai.assert;
+
+const assertTemplate = document.createElement('template');
+
+/**
+ * Removes comments, normalizes text content, and normalizes style attribute
+ * serialization.
+ *
+ * Although semantic-dom-diff removes comments, it does not normalize
+ * textContent after removing them, so differences in comments remain
+ * semantically meaningful; we currently use different comment marker strategies
+ * between client and server, so this inures us to those differences.
+ */
+const normalize = (node: Node | null) => {
+  if (node === null) {
+    return;
+  }
+  let nextSibling = node.nextSibling;
+  // Remove comments
+  if (node.nodeType === node.COMMENT_NODE) {
+    node.parentNode!.removeChild(node);
+  }
+  // Normalize text content (we don't use node.normalize, because
+  // it doesn't work on IE)
+  else if (node.nodeType === node.TEXT_NODE) {
+    while (
+      node.nextSibling &&
+      (node.nextSibling.nodeType === node.TEXT_NODE ||
+        node.nextSibling.nodeType === node.COMMENT_NODE)
+    ) {
+      if (node.nextSibling.nodeType === node.TEXT_NODE) {
+        node.nodeValue! += node.nextSibling.nodeValue;
+      }
+      node.parentNode!.removeChild(node.nextSibling);
+    }
+    nextSibling = node.nextSibling;
+  } else {
+    // When parsed via the parser, the browser maintains user-authored whitespace;
+    // when reset via the property it re-serializes in a normalized form
+    if (node.nodeType === node.ELEMENT_NODE) {
+      if ((node as HTMLElement).hasAttribute('style')) {
```

Comment:
Collapse condition into previous `if`?

Reply:
The comment here is unclear. Why reset `cssText`?

Reply:
to normalize the whitespace

Reply:
When parsed via the parser, the browser maintains user-authored whitespace: 
```
<div style="   color : red ; "></div>
```
```
div.getAttribute('style') === '   color : red ; ';
```
When reset via the property it re-serializes in a normalized form:
```
div.style.cssText = div.style.cssText;
```
```
div.getAttribute('style') === 'color: red;';
```

Reply:
Since the client-side code carries the `styleMap` logic to serialize the string for SSR, it would be extra code to make it exactly match the normalized form; however, when we do client-side updates via `styleMap` into the `style` object, the attribute will get re-serialized in a normalized form. So this allows us to write all expectations to the normalized form.

Reply:
ok